File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.cpp
#include<vector>
#include <string>
#include <queue>
using namespace std;
int dx[4]={1,-1,0,0};
int dy[4]={0,0,1,-1};
int dist[101][101];
int solution(vector<vector<int>> maps)
{
    int n=maps.size();
    int m=maps[0].size();
    queue<pair<int,int>> q;
    dist[0][0]=1;
    q.push({0,0});
    while(!q.empty())
    {
        int x=q.front().first;
        int y=q.front().second;
        q.pop();
        for(int i=0;i<4;i++)
        {
            int nx=x+dx[i];
            int ny=y+dy[i];
            if(nx>=0&&ny>=0&&nx<n&&ny<m&&maps[nx][ny]==1&&dist[nx][ny]==0)
            {
                dist[nx][ny]=dist[x][y]+1;
                q.push({nx,ny});
                
            }
        }
    }
    
    if(dist[n-1][m-1]==0) return -1;
    return dist[n-1][m-1];
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;
class Solution {
    public int solution(int[][] maps) {
        int answer = 0;
        
        int[] dx = {0,0,-1,1};
        int[] dy = {1,-1,0,0};
        Queue<int[]> q = new LinkedList<>();
        boolean[][] visited = new boolean[101][101];
        int n = maps.length;
        int m = maps[0].length;
        
        q.offer(new int[]{0,0,1});
        visited[0][0] = true;
        
        while(!q.isEmpty())
        {
            int[] cur = q.poll();
            int cx = cur[0];
            int cy = cur[1];
            int dist = cur[2];
            
            if(cx == n-1 && cy == m-1) {
                return dist;
            }
            
            for(int i = 0; i < 4; i++) {
                int nx = cx + dx[i];
                int ny = cy + dy[i];
                if(nx >= 0 && ny >= 0 && nx < n && ny < m) {
                    if(!visited[nx][ny] && maps[nx][ny] == 1) {
                        visited[nx][ny] = true;
                        q.offer(new int[]{nx, ny, dist + 1});
                    }
                }
            }
        }
        return -1;
    }
}